### PR TITLE
TOC aborts under service failure

### PR DESCRIPTION
**Pull request by jason-bragg**
Fixed TOC aborts under service failure
Added minimal failure testing for TOC.

### CODE REVIEW

File: src/Orleans.Transactions/State/TransactionQueue.cs [hunk 1]
```diff
@@ -482,6 +482,10 @@ private async Task StorageWork()
                     {
                         // process all committable entries, adding storage events to the storage batch
                         CollectEventsForBatch(committableEntries);
+                        if (problemFlag != TransactionalStatus.Ok)
```
Comment (ReubenBond) [inline src/Orleans.Transactions/State/TransactionQueue.cs]: `problemFlag` reminds me of [errno](http://man7.org/linux/man-pages/man3/errno.3.html) and [GetLastError](https://msdn.microsoft.com/en-us/library/windows/desktop/ms679360%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396), which makes me uncomfortable.

File: src/Orleans.Transactions/State/TransactionQueue.cs [hunk 2]
```diff
@@ -482,6 +482,10 @@ private async Task StorageWork()
...
+                        {
+                            return;
+                        }
```
Comment (ReubenBond) [inline src/Orleans.Transactions/State/TransactionQueue.cs]: Should this be calling `NotifyOfRestore()`?

**Comment by ReubenBond** [comment]
LGTM as long as the Notify comment is addressed

File: src/Orleans.Transactions/State/TransactionQueue.cs (hunk 1 above)
Comment (jason-bragg) [reply src/Orleans.Transactions/State/TransactionQueue.cs]: Agree.  I openly acknowledge that this solution is kludgy, almost to the point of being a hack.  All three loops (lock worker, storage worker, and confirmation worker) need refactored.  I'm trying to fix things within the current code until I've time to do refactoring passes.

File: src/Orleans.Transactions/State/TransactionQueue.cs (hunk 2 above)
Comment (jason-bragg) [reply src/Orleans.Transactions/State/TransactionQueue.cs]: No.  The finally block will check the problemFlag and handle any issues.  If Notify needs be called again, the finally block will do so.

**Pull request merged**